### PR TITLE
CDAP-5186 add data pipeline application

### PR DESCRIPTION
**Pull request by albertshau**
Add the application for data pipeline. The app will use the
pipeline planner to break up a logical pipeline into physical
phases. Each phase is a program with its own source(s), transforms,
and sink(s). Phases are connected through local datasets.

### CODE REVIEW

**Comment by albertshau** [comment]
http://builds.cask.co/browse/CDAP-DUT3737

File: cdap-app-templates/cdap-etl/cdap-etl-data-pipeline/src/test/java/co/cask/cdap/etl/datapipeline/DataPipelineTest.java [hunk 1]
```diff
@@ -0,0 +1,138 @@
...
+  public static final TestConfiguration CONFIG = new TestConfiguration("explore.enabled", false);
+
+  protected static final Id.Artifact APP_ARTIFACT_ID = Id.Artifact.from(Id.Namespace.DEFAULT, "datapipeline", "3.4.0");
+  protected static final ArtifactSummary ETLBATCH_ARTIFACT = new ArtifactSummary("datapipeline", "3.4.0");
+
+  private static int startCount;
+
+  @BeforeClass
+  public static void setupTest() throws Exception {
+    if (startCount++ > 0) {
```
Comment (anwar6953) [inline cdap-app-templates/cdap-etl/cdap-etl-data-pipeline/src/test/java/co/cask/cdap/etl/datapipeline/DataPipelineTest.java]: Why is this needed? This method is a `@BeforeClass`.

Comment (albertshau) [inline cdap-app-templates/cdap-etl/cdap-etl-data-pipeline/src/test/java/co/cask/cdap/etl/datapipeline/DataPipelineTest.java]: oops, from copy-paste. will remove

File: cdap-app-templates/cdap-etl/cdap-etl-data-pipeline/src/main/java/co/cask/cdap/etl/datapipeline/SmartWorkflow.java [hunk 2]
```diff
@@ -0,0 +1,162 @@
...
+    addProgram(node, new TrunkProgramAdder(configurer));
+
+    Set<String> outputs = dag.getNodeOutputs(node);
+    if (outputs.isEmpty()) {
+      return;
+    }
+
+    // if this is a fork
+    if (outputs.size() > 1) {
+      WorkflowForkConfigurer<? extends WorkflowConfigurer> forkConfigurer = configurer.fork();
```
Comment (sreevatsanraman) [inline cdap-app-templates/cdap-etl/cdap-etl-data-pipeline/src/main/java/co/cask/cdap/etl/datapipeline/SmartWorkflow.java]: There is no join for fork.

Comment (sreevatsanraman) [inline cdap-app-templates/cdap-etl/cdap-etl-data-pipeline/src/main/java/co/cask/cdap/etl/datapipeline/SmartWorkflow.java]: nvm. see it below.

**Comment by sreevatsanraman** [comment]
LGTM. Merge after Ali's comments are addressed.

**Comment by albertshau** [comment]
fixed and squashed. will merge after a green build
